### ISSUE TITLE
conj is not tagged with __DEVICE__

### ISSUE THREAD

**Issue by aminiussi** [llvm/llvm-project#113207]
In 'clang/lib/Headers/openmp_wrappers/complex_cmath.h', conj is not tagged with `__DEVICE__`. As a result it is not marcked constexpr (as other functions in that header).

Now, that file is included by 'clang/lib/Headers/openmp_wrappers/complex' which in turn includes the platform complex header, which can/will include gcc complex header. 

But it seems that with gcc-12 and current version, [it is constexpr](https://github.com/gcc-mirror/gcc/blob/cb131a401b7489cc17e2d70420cf9a916515b3f6/libstdc%2B%2B-v3/include/std/complex#L91) (or at least as constepxr as 'norm', which is tagged '__DEVICE__')

So there seems to be an inconsistency which results in:

```
In file included from /opt/cray/pe/cce/17.0.0/cce-clang/x86_64/lib/clang/17/include/openmp_wrappers/complex:51:
/opt/cray/pe/cce/17.0.0/cce-clang/x86_64/lib/clang/17/include/openmp_wrappers/complex_cmath.h:68:40: error: non-constexpr declaration of 'conj' follows constexpr declaration
   68 | template <class _Tp> std::complex<_Tp> conj(const std::complex<_Tp> &__c) {
      |                                        ^
/opt/rh/gcc-toolset-12/root/usr/lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/complex:709:5: note: previous declaration is here
  709 |     conj(const complex<_Tp>& __z)
      |     ^
1 error generated.
```

The problem arise with Cray's flavor of clang++ (which is a clang++ to which they add bugs apparently):

```
21:15:49 [alainm@login4 rel]#CC --version
Cray clang version 17.0.0  (b59b7a8e9169719529cf5ab440f3c301e515d047)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/cray/pe/cce/17.0.0/cce-clang/x86_64/share/../bin
21:17:13 [alainm@login4 rel]#
```

So I understand that Cray issues are their issues but header wise, there seems to be an inconsistency since, if llvm headers includes the gnu headers, the constexpr modifiers of complex function should follow, right ?

thanks

**Comment by dmikushin** [comment]
I would like to confirm that this issue is not Cray-specific, and takes place in the mainline Clang 17:

```
In file included from /usr/bin/../lib/clang/17/include/openmp_wrappers/complex:51:
/usr/bin/../lib/clang/17/include/openmp_wrappers/complex_cmath.h:68:40: error: non-constexpr declaration of 'conj' follows constexpr declaration
   68 | template <class _Tp> std::complex<_Tp> conj(const std::complex<_Tp> &__c) {
      |                                        ^
/opt/rh/gcc-toolset-13/root/usr/lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/complex:970:5: note: previous declaration is here
  970 |     conj(const complex<_Tp>& __z)
      |     ^
```

**Comment by dmikushin** [comment]
A minimal example can be cooked simply by using `#include <complex>` together with `-fopenmp -fopenmp-targets=nvptx64-nvidia-cuda` flags.

**Comment by chandraghale** [comment]
@aminiussi  @dmikushin 
I observed that the test code  runs successfully with GNU++17 but encounters issues with GNU++20. This seems to be C++20's P0415 update [https://github.com/RTEMS/gnu-mirror-gcc/commit/e987fb1ebecc7494f9869071e487deef70a792f6](url) which made the `conj` function in `<complex>` `constexpr`. During OpenMP offloading, `complex_cmath.h` is included, where `conj` is not marked as `constexpr`, leading to a compilation failure since the forward declaration in `<complex>` does specify `constexpr`. To resolve this, I propose if we can update  the OpenMP wrapper header `complex_cmath.h`  to align with C++20's `constexpr` specification for `conj`, which should address this incompatibility ? 

Tested with this change ,it is working. Defined new macro CXX20_CONSTEXPR_DEVICE to make conditionally constexpr under C++20 while remaining compatible with earlier standards.
```

--- a/clang/lib/Headers/openmp_wrappers/complex_cmath.h
+++ b/clang/lib/Headers/openmp_wrappers/complex_cmath.h
@@ -64,8 +64,12 @@ template <class _Tp> __DEVICE__ _Tp norm(const std::complex<_Tp> &__c) {
 }

 // conj
-
-template <class _Tp> std::complex<_Tp> conj(const std::complex<_Tp> &__c) {
+#ifdef  _GLIBCXX20_CONSTEXPR
+#define CXX20_CONSTEXPR_DEVICE  __DEVICE__
+#else
+#define CXX20_CONSTEXPR_DEVICE
+#endif
+template <class _Tp> CXX20_CONSTEXPR_DEVICE  std::complex<_Tp> conj(const std::complex<_Tp> &__c) {
   return std::complex<_Tp>(__c.real(), -__c.imag());
 }

```

**Comment by soumitrachatterjee** [comment]
This is because OpenMP uses wrappers around the complex math functions, which in turn use the C++ standard library headers. And unfortunately, the implementation of the `conj` function is not correctly defined in the wrappers to be compatible with C++20.

From the standard library header, we see the correct declaration of `conj`:
///  Return complex conjugate of @a z.
template<typename _Tp>
  _GLIBCXX20_CONSTEXPR complex<_Tp> conj(const complex<_Tp>&);

Whereas the declaration in the OpenMP wrapper misses out on the C++20 `constexpr`:
// conj
template <class _Tp> std::complex<_Tp> conj(const std::complex<_Tp> &__c) {
  return std::complex<_Tp>(__c.real(), -__c.imag());
}

@chandraghale I believe the correct fix would be to simply use the macro `_GLIBCXX20_CONSTEXPR`?

**Comment by soumitrachatterjee** [comment]
> The problem arise with Cray's flavor of clang++ (which is a clang++ to which they add bugs apparently):
No, we don't add bugs - we only fix them :)

**Comment by chandraghale** [comment]
> This is because OpenMP uses wrappers around the complex math functions, which in turn use the C++ standard library headers. And unfortunately, the implementation of the `conj` function is not correctly defined in the wrappers to be compatible with C++20.
> 
> From the standard library header, we see the correct declaration of `conj`: /// Return complex conjugate of @A z. template _GLIBCXX20_CONSTEXPR complex<_Tp> conj(const complex<_Tp>&);
> 
> Whereas the declaration in the OpenMP wrapper misses out on the C++20 `constexpr`: // conj template std::complex<_Tp> conj(const std::complex<_Tp> &__c) { return std::complex<_Tp>(__c.real(), -__c.imag()); }
> 
> @chandraghale I believe the correct fix would be to simply use the macro `_GLIBCXX20_CONSTEXPR`?

Thanks @soumitrachatterjee  for explanation. 
 The `__DEVICE__` macro is used to improve compatibility with GPU offloading environments, instead of using only  `_GLIBCXX20_CONSTEXPR`. The `__DEVICE__` macro expands to `static constexpr __attribute__((nothrow))`, which provides internal linkage and guarantees that functions are exception-free. This looks better suited for GPU environments.

**Comment by dmikushin** [comment]
> This is because OpenMP uses wrappers around the complex math functions, which in turn use the C++ standard library headers. And unfortunately, the implementation of the `conj` function is not correctly defined in the wrappers to be compatible with C++20.

@soumitrachatterjee I'm wondering why it is designed this way. First of all, why OpenMP needs to care about the complex math functions? I guess, it has something to do with the OpenMP reduction support. But the most important question is: **why in LLVM this support is provided in the form of potentially breaking C++ headers, not in the form of LLVM IR?** Someone who decided this must share their motivation with us.

**Comment by chandraghale** [comment]
@dmikushin  @soumitrachatterjee  I agree with your comments and concerns. From previous patches, I see that OpenMP-compatible versions of these complex math functions are created by overloading the standard library's template functions, still relying on standard headers. This approach indeed has the potential to cause conflicts with C++ headers. For now, I am following the existing approach and raising the PR [https://github.com/llvm/llvm-project/pull/115306](url) to help unblocking  this current  issue.

**Comment by dmikushin** [comment]
@chandraghale Thank you for creating a PR. Kindly, could you please also start a discussion on LLVM Discourse to raise a concern about the future portability issues of this design?
